02-peer-protocol.md

```diff
@@ -938,6 +960,38 @@ previously sent.
 A receiving node MAY ignore spurious message retransmission, or MAY
 fail the channel if they occur.
 
+### Rationale
+
+The effect of requirements above are that the opening phase is almost
+atomic: if it doesn't complete, it starts again.  The only exception
+is where the `funding_signed` message is sent and not received: in
+this case, the funder will forget the channel and presumably open
+a new one on reconnect; the other node will eventually forget the
+original channel due to never receiving `funding_locked`.
+
+For normal operation and during shutdown, the updates are similarly
+forgotten, though in this case the loss of `revoke_and_ack` can cause
+an entire retransmission of the previous updates and commitment:
+this is OK, as the recipient will simply ignore duplicates (as
+documented in the requirements for [`update_add_htlc`](#adding-an-htlc-update_add_htlc) and [Removing an HTLC](#removing-an-htlc-update_fulfill_htlc-update_fail_htlc-and-update_fail_malformed_htlc).
+
+This retransmission is required to generally be a super-set of the
+original, to give the same result whether the first transmission was
+committed or not.  The exception for `update_fee` applies because
+`update_fee` simply overrides previous `update_fee` messages, though a
+retransmission has to send at least one `update_fee` message if one was
+sent originally.
+
+Note that if no `commitment_signed` was sent in the first place,
+there's no requirement to retransmit those (uncommitted) updates.
```

Comment:
If you agree on that, I don't think we can allow the mutation of previously sent `update_fee`, because a fee that the funder cannot afford may invalidate subsequent `update_add_htlc`, right? Even if the sender can afford a new higher fee, the fact that we break the "simple replay" scenario kind of defeats the purpose in my opinion.

To be clear, I see two options, either:
(1) we only allow a single flow of updates (add, fail, fulfill, fee), and reconnection is a simple replay: simpler and maybe best for now?
or 
(2) we allow forking the update flow in case of a disconnection, and reconnection is a rollback: slightly more complex, but maybe more optimal?

I think all updates should be treated the same way, and if we already know we will introduce an `update_unadd_htlc`, then I think (1) might be the way to go. I may be influenced by having worked on database wal!